pkg/services/cost/analyzers/snowflake/application.go

```diff
@@ -0,0 +1,208 @@
+package snowflake
+
+import (
+	"database/sql"
+	"fmt"
+	"github.com/de-tools/data-atlas/pkg/models/domain"
+	"time"
+)
+
+type applicationAnalyzer struct {
+	db *sql.DB
+}
+
+func NewApplicationAnalyzer(db *sql.DB) *applicationAnalyzer {
+	return &applicationAnalyzer{
+		db: db,
+	}
+}
+
+func (aa *applicationAnalyzer) GetResourceType() string {
+	return "application"
+}
+
+func (aa *applicationAnalyzer) CollectUsage(days int) ([]domain.ResourceCost, error) {
+	query := `
+		SELECT 
+			application_id,
+			application_name,
+			credits_used,
+			credits_used_breakdown,
+			listing_global_name,
+			storage_bytes,
+			storage_bytes_breakdown,
+			usage_date
+		FROM snowflake.account_usage.application_daily_usage_history
+		WHERE usage_date >= dateadd(day, ?, current_date())
+	`
+
+	rows, err := aa.db.Query(query, -days)
+	if err != nil {
+		return nil, fmt.Errorf("application usage query failed: %w", err)
+	}
+	defer rows.Close()
+
+	var costs []domain.ResourceCost
+	for rows.Next() {
+		var appID, appName, globalName string
+		var credits float64
+		var storageBytes int64
+		var creditsBreakdown, storageBreakdown string
+		var usageDate time.Time
+
+		if err := rows.Scan(&appID, &appName, &credits, &creditsBreakdown, &globalName,
+			&storageBytes, &storageBreakdown, &usageDate); err != nil {
+			return nil, err
+		}
+
+		cost := domain.ResourceCost{
+			StartTime: usageDate,
+			EndTime:   usageDate.Add(24 * time.Hour),
+			Resource: domain.Resource{
+				Platform:    "Snowflake",
+				Service:     "Application",
+				Name:        appName,
+				Description: fmt.Sprintf("Snowflake Application %s (%s)", appName, globalName),
+				Tags: map[string]string{
+					"global_name": globalName,
+				},
+				Metadata: struct {
+					ID        string
+					AccountID string
+					UserID    string
+					Region    string
+				}{
+					ID: appID,
+				},
+			},
+			Costs: []domain.CostComponent{
+				{
+					Type:        "compute",
+					Value:       credits,
+					Unit:        "credits",
+					Rate:        3.0,
+					TotalAmount: credits * 3,
+					Currency:    "USD",
+					Description: "Application compute credits",
+				},
+				{
+					Type:        "storage",
+					Value:       float64(storageBytes),
+					Unit:        "bytes",
+					Rate:        40.0 / (1024 * 1024 * 1024 * 1024), // $40 per TB
+					TotalAmount: float64(storageBytes) * 40 / (1024 * 1024 * 1024 * 1024),
+					Currency:    "USD",
+					Description: "Application storage",
+				},
+			},
+		}
+		costs = append(costs, cost)
+	}
+	return costs, nil
+}
+
+func (aa *applicationAnalyzer) GenerateReport(days int) (*domain.Report, error) {
```

Comment:
Does it make sense to extract report generation into a separate struct? It seems that collecting and representing the data can be two separate concerns

Reply:
report currently is specific to analyzer, while it's possible to abstract it away this would bring potential mistakes.
I will leave it for now. Perhaps we'd need to extend the domain and make report into cost models (part of it).